# Homework

**_NOTE:_** sometimes your answer doesn't match one of the options exactly. That's fine. Select the option that's closest to your solution.

## Dataset

In this homework, we will use the lead scoring dataset Bank Marketing dataset. Download it from here.

Or you can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
```

In this dataset our desired target for classification task will be converted variable - has the client signed up to the platform or not.

## Data preparation

- Check if the missing values are presented in the features.
- If there are missing values:
  - For caterogiral features, replace them with 'NA'
  - For numerical features, replace with with 0.0

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

In [2]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv'

In [3]:
!wget $data -O course_lead_scoring.csv

--2025-10-14 11:47:05--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
connected. to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... 
200 OKequest sent, awaiting response... 
Length: 80876 (79K) [text/plain]
Saving to: ‘course_lead_scoring.csv’

course_lead_scoring 100%[===================>]  78.98K  --.-KB/s    in 0.02s   

2025-10-14 11:47:05 (3.47 MB/s) - ‘course_lead_scoring.csv’ saved [80876/80876]



In [4]:
df = pd.read_csv('course_lead_scoring.csv')
df#.head()

lead_source       industry  number_of_courses_viewed  annual_income  \
0           paid_ads            NaN                         1        79450.0   
1       social_media         retail                         1        46992.0   
2             events     healthcare                         5        78796.0   
3           paid_ads         retail                         2        83843.0   
4           referral      education                         3        85012.0   
...              ...            ...                       ...            ...   
1457        referral  manufacturing                         1            NaN   
1458        referral     technology                         3        65259.0   
1459        paid_ads     technology                         1        45688.0   
1460        referral            NaN                         5        71016.0   
1461  organic_search        finance                         3        92855.0   

     employment_status       location  interaction_count  lead_score  \
0           unemployed  south_america                  4        0.94   
1             employed  south_america                  1        0.80   
2           unemployed      australia                  3        0.69   
3                  NaN      australia                  1        0.87   
4        self_employed         europe                  3        0.62   
...                ...            ...                ...         ...   
1457     self_employed  north_america                  4        0.53   
1458           student         europe                  2        0.24   
1459           student  north_america                  3        0.02   
1460     self_employed  north_america                  0        0.25   
1461           student  north_america                  3        0.41   

      converted  
0             1  
1             0  
2             1  
3             0  
4             1  
...         ...  
1457          1  
1458          1  
1459          1  
1460          1  
1461          1  

[1462 rows x 9 columns]

In [5]:
df.dtypes

lead_source                  object
industry                     object
number_of_courses_viewed      int64
annual_income               float64
employment_status            object
location                     object
interaction_count             int64
lead_score                  float64
converted                     int64
dtype: object

In [6]:
numerical = ['number_of_courses_viewed', 'annual_income', 'interaction_count', 'lead_score']
numerical = df.select_dtypes(exclude='object').columns.drop('converted')
numerical = df.select_dtypes(exclude='object').columns.drop('converted')
numerical

Index(['number_of_courses_viewed', 'annual_income', 'interaction_count',
       'lead_score'],
      dtype='object')

In [7]:
#categorical = ['lead_source', 'industry', 'employment_status', 'location']
categorical = df.select_dtypes(include=['object']).columns

In [8]:
categorical

Index(['lead_source', 'industry', 'employment_status', 'location'], dtype='object')

In [9]:
#df[categorical]
#df[['lead_source', 'industry', 'employment_status', 'location']]

In [10]:
df[numerical] = df[numerical].fillna(0.0)
df[categorical] = df[categorical].fillna('NA')

In [11]:
df

lead_source       industry  number_of_courses_viewed  annual_income  \
0           paid_ads             NA                         1        79450.0   
1       social_media         retail                         1        46992.0   
2             events     healthcare                         5        78796.0   
3           paid_ads         retail                         2        83843.0   
4           referral      education                         3        85012.0   
...              ...            ...                       ...            ...   
1457        referral  manufacturing                         1            0.0   
1458        referral     technology                         3        65259.0   
1459        paid_ads     technology                         1        45688.0   
1460        referral             NA                         5        71016.0   
1461  organic_search        finance                         3        92855.0   

     employment_status       location  interaction_count  lead_score  \
0           unemployed  south_america                  4        0.94   
1             employed  south_america                  1        0.80   
2           unemployed      australia                  3        0.69   
3                   NA      australia                  1        0.87   
4        self_employed         europe                  3        0.62   
...                ...            ...                ...         ...   
1457     self_employed  north_america                  4        0.53   
1458           student         europe                  2        0.24   
1459           student  north_america                  3        0.02   
1460     self_employed  north_america                  0        0.25   
1461           student  north_america                  3        0.41   

      converted  
0             1  
1             0  
2             1  
3             0  
4             1  
...         ...  
1457          1  
1458          1  
1459          1  
1460          1  
1461          1  

[1462 rows x 9 columns]

# Question 1

What is the most frequent observation (mode) for the column industry?

- `NA`
- `technology`
- `healthcare`
- `retail`

In [12]:
df.industry.mode()

0    retail
Name: industry, dtype: object

# Question 2

Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

- `interaction_count` and `lead_score`
- `number_of_courses_viewed` and `lead_score`
- `number_of_courses_viewed` and `interaction_count`
- `annual_income` and `interaction_count`

Only consider the pairs above when answering this question.
Split the data

- Split your data in train/val/test sets with 60%/20%/20% distribution.
- Use Scikit-Learn for that (the `train_test_split` function) and set the seed to ''42''.
- Make sure that the target value `y` is not in your dataframe.

In [13]:
df[numerical].corr()

number_of_courses_viewed  annual_income  \
number_of_courses_viewed                  1.000000       0.009770   
annual_income                             0.009770       1.000000   
interaction_count                        -0.023565       0.027036   
lead_score                               -0.004879       0.015610   

                          interaction_count  lead_score  
number_of_courses_viewed          -0.023565   -0.004879  
annual_income                      0.027036    0.015610  
interaction_count                  1.000000    0.009888  
lead_score                         0.009888    1.000000

interaction_count and lead_score: 0.009888, number_of_courses_viewed and lead_score: -0.004879

number_of_courses_viewed and interaction_count: -0.023565


annual_income and interaction_count: 0.027036


In [14]:
from sklearn.model_selection import train_test_split

In [15]:
df_train_full, df_test = train_test_split(df.drop(columns='converted'), test_size=0.2, random_state=42)

In [16]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [17]:
df_train_full

lead_source       industry  number_of_courses_viewed  annual_income  \
1066  social_media  manufacturing                         2        44403.0   
638         events         retail                         3        38048.0   
799   social_media      education                         2        71399.0   
380       referral      education                         2        47912.0   
303       paid_ads     healthcare                         1        34806.0   
...            ...            ...                       ...            ...   
1130        events  manufacturing                         1        57039.0   
1294        events     healthcare                         2        56185.0   
860       paid_ads  manufacturing                         1        56402.0   
1459      paid_ads     technology                         1        45688.0   
1126        events  manufacturing                         0        43862.0   

     employment_status       location  interaction_count  lead_score  
1066     self_employed      australia                  1        0.71  
638            student  north_america                  6        0.97  
799                 NA         europe                  1        0.51  
380           employed      australia                  1        0.04  
303           employed         europe                  4        0.32  
...                ...            ...                ...         ...  
1130          employed  south_america                  2        0.30  
1294           student  south_america                  2        0.44  
860            student  north_america                  1        0.02  
1459           student  north_america                  3        0.02  
1126     self_employed           asia                  4        0.21  

[1169 rows x 8 columns]

In [18]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [19]:
y_train = df_train.values
y_val = df_val.values
y_test = df_test.values

In [20]:
df_train_full = df_train_full.reset_index(drop=True)

# Question 3

- Calculate the mutual information score between y and other categorical variables in the dataset. Use the training set only.
- Round the scores to 2 decimals using round(score, 2).

Which of these variables has the biggest mutual information score?

- `industry`
- `location`
- `lead_source`
- `employment_status`

In [21]:
from sklearn.metrics import mutual_info_score

In [22]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=42)

In [23]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [24]:
def calculate_mi(series):
    return mutual_info_score(series, df_train.converted)

df_mi = df_train[categorical].apply(calculate_mi)
df_mi = df_mi.sort_values(ascending=False).to_frame(name='MI')


df_mi.round(2)

MI
lead_source        0.04
employment_status  0.01
industry           0.01
location           0.00

lead_source

In [25]:
#df_train_full, df_test = train_test_split(df.drop(columns='converted'), test_size=0.2, random_state=42)

# Question 4

- Now let's train a logistic regression.
- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
- To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
- model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
- Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

What accuracy did you get?

- 0.64
- 0.74
- 0.84
- 0.94

In [26]:
## One-hot encodingabs

In [27]:
from sklearn.feature_extraction import DictVectorizer

In [28]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [29]:
df_train

lead_source       industry  number_of_courses_viewed  annual_income  \
1077        paid_ads         retail                         0        58472.0   
463   organic_search  manufacturing                         3        71738.0   
842         paid_ads     technology                         3        81973.0   
835               NA     technology                         1        74956.0   
837   organic_search         retail                         3        59335.0   
...              ...            ...                       ...            ...   
725   organic_search          other                         1        43907.0   
401     social_media         retail                         3        64969.0   
957               NA      education                         3        89042.0   
992     social_media  manufacturing                         1            0.0   
1267          events        finance                         4        50259.0   

     employment_status       location  interaction_count  lead_score  \
1077           student    middle_east                  5        0.03   
463            student    middle_east                  6        0.77   
842           employed  north_america                  2        0.59   
835           employed         europe                  3        0.34   
837            student      australia                  1        0.98   
...                ...            ...                ...         ...   
725           employed      australia                  4        0.33   
401           employed  north_america                  1        0.18   
957           employed           asia                  4        0.75   
992      self_employed         europe                  1        0.65   
1267           student           asia                  3        0.54   

      converted  
1077          0  
463           1  
842           1  
835           1  
837           1  
...         ...  
725           1  
401           0  
957           1  
992           0  
1267          1  

[876 rows x 9 columns]

In [30]:
categorical = list(categorical)
numerical = list(numerical)

In [31]:
train_dict = df_train[categorical + numerical].to_dict(orient='records')
train_dict[0]

{'lead_source': 'paid_ads',
 'industry': 'retail',
 'employment_status': 'student',
 'location': 'middle_east',
 'number_of_courses_viewed': 0,
 'annual_income': 58472.0,
 'interaction_count': 5,
 'lead_score': 0.03}

In [32]:
dv = DictVectorizer(sparse=False)
train_dict = df_train[categorical + numerical].to_dict(orient='records')
X_train = dv.fit_transform(train_dict)

val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

In [33]:
X_train = dv.fit_transform(train_dict)

In [34]:
X_train.shape

(876, 31)

In [35]:
dv.get_feature_names_out()

array(['annual_income', 'employment_status=NA',
       'employment_status=employed', 'employment_status=self_employed',
       'employment_status=student', 'employment_status=unemployed',
       'industry=NA', 'industry=education', 'industry=finance',
       'industry=healthcare', 'industry=manufacturing', 'industry=other',
       'industry=retail', 'industry=technology', 'interaction_count',
       'lead_score', 'lead_source=NA', 'lead_source=events',
       'lead_source=organic_search', 'lead_source=paid_ads',
       'lead_source=referral', 'lead_source=social_media', 'location=NA',
       'location=africa', 'location=asia', 'location=australia',
       'location=europe', 'location=middle_east',
       'location=north_america', 'location=south_america',
       'number_of_courses_viewed'], dtype=object)

In [36]:
y_train = df_train.converted.values
y_val = df_val.converted.values

In [37]:
from sklearn.linear_model import LogisticRegression

In [38]:
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [39]:
model.predict_proba(X_val)

array([[0.38807837, 0.61192163],
       [0.20017383, 0.79982617],
       [0.46978656, 0.53021344],
       [0.52868521, 0.47131479],
       [0.42933868, 0.57066132],
       [0.55772831, 0.44227169],
       [0.12872331, 0.87127669],
       [0.15116885, 0.84883115],
       [0.16709963, 0.83290037],
       [0.38502199, 0.61497801],
       [0.45031973, 0.54968027],
       [0.21846912, 0.78153088],
       [0.30960214, 0.69039786],
       [0.22982878, 0.77017122],
       [0.4734056 , 0.5265944 ],
       [0.08293575, 0.91706425],
       [0.46829365, 0.53170635],
       [0.57876951, 0.42123049],
       [0.69853545, 0.30146455],
       [0.15118417, 0.84881583],
       [0.20511347, 0.79488653],
       [0.26329625, 0.73670375],
       [0.55472789, 0.44527211],
       [0.35161617, 0.64838383],
       [0.5823118 , 0.4176882 ],
       [0.24606582, 0.75393418],
       [0.09833884, 0.90166116],
       [0.66096951, 0.33903049],
       [0.56818569, 0.43181431],
       [0.0319319 , 0.9680681 ],
       [0.

In [40]:
y_pred = model.predict_proba(X_val)[:, 1]

In [41]:
converted = y_pred >= 0.5

In [42]:
y_pred.mean()

0.6587082147463391

In [43]:
(y_val == converted).mean()

0.6996587030716723

In [44]:
y_pred = model.predict(X_val)

In [45]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [46]:
round(accuracy_score(y_val, y_pred),2)

0.7

In [47]:
dicts_train_full = df_train_full[categorical + numerical].to_dict(orient='records')

In [48]:
dv = DictVectorizer(sparse=False)

In [49]:
X_train_full = dv.fit_transform(dicts_train_full)

In [50]:
y_train_full = df_train_full.converted.values

In [51]:
model_baseline = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model_baseline.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [52]:
model.intercept_[0]

-0.0691472802783609

In [53]:
model.coef_[0].round(3)

array([-0.   , -0.015,  0.034,  0.003,  0.012, -0.103, -0.025,  0.049,
       -0.02 , -0.013, -0.003, -0.009, -0.032, -0.016,  0.311,  0.051,
        0.02 , -0.012, -0.012, -0.115,  0.08 , -0.03 ,  0.004, -0.011,
       -0.011, -0.006,  0.008,  0.006, -0.033, -0.025,  0.454])

In [54]:
dicts_test = df_test[categorical + numerical].to_dict(orient='records')

In [55]:
X_test = dv.transform(dicts_test)

In [56]:
y_pred = model.predict_proba(X_train)[:, 1]

In [57]:
converted_decision = (y_pred >= 0.5)
converted_decision

array([ True,  True,  True,  True,  True,  True,  True,  True, False,
        True, False,  True, False, False,  True,  True,  True,  True,
        True, False,  True,  True,  True, False, False, False,  True,
        True,  True,  True, False,  True, False,  True,  True,  True,
        True,  True, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True, False,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True, False,
        True, False,  True,  True,  True,  True,  True,  True, False,
        True,  True,  True, False,  True,  True,  True, False,  True,
        True,  True, False, False,  True, False,  True,  True, False,
        True,  True, False,  True, False,  True, False,  True,  True,
       False, False,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True,  True,
       False,  True,

In [58]:
#(y_val == converted_decision).mean()

In [59]:
y_pred = model.predict(X_val)

In [60]:
accuracy_score(y_val, y_pred)

0.6996587030716723

In [61]:
ac = accuracy_score(y_val, y_pred)

In [62]:
y_val_pred_baseline = model_baseline.predict(X_val)
accuracy_baseline = accuracy_score(y_val, y_val_pred_baseline)
print(f'Accuracy: {accuracy_baseline:.2f}')

Accuracy: 0.70


In [63]:
round(ac,2)

0.7

In [64]:
# A function to evaluate the model without a specific feature
def evaluate_without_feature(feature_to_exclude):
    # Get the feature names from the DictVectorizer
    feature_names = dv.get_feature_names_out()
    
    # Find indices of features to keep (all except the ones to exclude)
    exclude_indices = [i for i, feat in enumerate(feature_names) if feature_to_exclude in feat]
    keep_indices = [i for i in range(len(feature_names)) if i not in exclude_indices]
    
    # Create new datasets without the feature
    X_train_reduced = X_train[:, keep_indices]
    X_val_reduced = X_val[:, keep_indices]
    
    # Train the model
    model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
    model.fit(X_train_reduced, y_train)
    
    # Evaluate on validation set
    y_val_pred = model.predict(X_val_reduced)
    accuracy = accuracy_score(y_val, y_val_pred)
    
    return accuracy

# Question 5

- Let's find the least useful feature using the feature elimination technique.
- Train a model using the same features and parameters as in Q4 (without rounding).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

- `'industry'`
- `'employment_status'`
- `'lead_score'`

**_NOTE:_** The difference doesn't have to be positive.

numerical = ['number_of_courses_viewed', 'annual_income', 'interaction_count', 'lead_score']
categorical = ['lead_source', 'industry', 'employment_status', 'location']

In [65]:
df[['lead_source', 'converted']]

lead_source  converted
0           paid_ads          1
1       social_media          0
2             events          1
3           paid_ads          0
4           referral          1
...              ...        ...
1457        referral          1
1458        referral          1
1459        paid_ads          1
1460        referral          1
1461  organic_search          1

[1462 rows x 2 columns]

In [66]:
#df_train_full, df_test = train_test_split(df.drop(columns=['converted','lead_source']), test_size=0.2, random_state=42)

In [67]:
X = df[numerical + categorical]
y = df['converted']

In [68]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=42)

In [69]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [70]:
dicts_train_full = df_train_full[categorical + numerical].to_dict(orient='records')

In [71]:
dv = DictVectorizer(sparse=False)
X_train_full = dv.fit_transform(dicts_train_full)

In [72]:
y_train_full = df_train_full.converted.values

In [73]:
dicts_test = df_test[categorical + numerical].to_dict(orient='records')

In [74]:
# Evaluate the model without each feature
features_to_test = ['industry', 'employment_status', 'lead_score']
results = {}

for feature in features_to_test:
    accuracy_without_feature = evaluate_without_feature(feature)
    difference = accuracy_baseline - accuracy_without_feature
    results[feature] = {
        'accuracy': accuracy_without_feature,
        'difference': difference
    }
    print(f"Without {feature}: Accuracy = {accuracy_without_feature:.6f}, Difference = {difference:.6f}")

# Find the feature with the smallest difference
min_feature = min(results, key=lambda x: abs(results[x]['difference']))
print(f"\nFeature with smallest impact: {min_feature}, Difference: {results[min_feature]['difference']:.6f}")

Without industry: Accuracy = 0.699659, Difference = 0.000000
Without employment_status: Accuracy = 0.696246, Difference = 0.003413
Without lead_score: Accuracy = 0.706485, Difference = -0.006826

Feature with smallest impact: industry, Difference: 0.000000


# Question 6

- Now let's train a regularized logistic regression.
- Let's try the following values of the parameter `C: [0.01, 0.1, 1, 10, 100]`.
- Train models using all the features as in Q4.
- Calculate the accuracy on the validation dataset and round it to 3 decimal digits.

Which of these `C` leads to the best accuracy on the validation set?

- 0.01
- 0.1
- 1
- 10
- 100

**_NOTE:_** If there are multiple options, select the smallest C.

In [76]:

c_values = [0.01, 0.1, 1, 10, 100]

# Dictionary to store results
c_results = {}

# Train and evaluate models for each C value
for c in c_values:
    # Train model with current C value
    model = LogisticRegression(solver='liblinear', C=c, max_iter=1000, random_state=42)
    model.fit(X_train, y_train)
    
    # Evaluate on validation set
    y_val_pred = model.predict(X_val)
    accuracy = accuracy_score(y_val, y_val_pred)
    
    # Store result
    c_results[c] = round(accuracy, 3)
    
    print(f"C = {c}, Validation Accuracy = {c_results[c]}")

best_accuracy = max(c_results.values())
best_c_values = [c for c, acc in c_results.items() if acc == best_accuracy]
best_c = min(best_c_values)

print(f"\nBest C value: {best_c} with validation accuracy: {c_results[best_c]}")

C = 0.01, Validation Accuracy = 0.7
C = 0.1, Validation Accuracy = 0.7
C = 1, Validation Accuracy = 0.7
C = 10, Validation Accuracy = 0.7
C = 100, Validation Accuracy = 0.7

Best C value: 0.01 with validation accuracy: 0.7


# Submit the results

- Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2025/homework/hw03
- If your answer doesn't match options exactly, select the closest one